### PR TITLE
Since SynchronizedCollection is only used in one place rename it to f…

### PR DESCRIPTION
**Pull request by Mpdreamz**
…it its purpose, also exposes an AddRange for adding many documents at once.

This also makes `Items` on `BulkResponse` `IReadOnlyList<T>` to allow indexed access.

### CODE REVIEW

File: src/Nest/Document/Multiple/Bulk/BulkRequest.cs [hunk 1]
```diff
@@ -134,10 +141,11 @@ public BulkDescriptor IndexMany<T>(IEnumerable<T> @objects, Func<BulkIndexDescri
 		) where T : class
 		  where TPartialDocument : class
 		{
-			foreach (var item in @objects)
-				AddOperation(bulkUpdateSelector.InvokeOrDefault(new BulkUpdateDescriptor<T, TPartialDocument>().IdFrom(item), item));
-
-			return this;
+			var operations = @objects
```
Comment (Mpdreamz) [inline src/Nest/Document/Multiple/Bulk/BulkRequest.cs]: Talked about this and we'll rewrite to `foreach` loop or consume Enumerator directly

File: src/Nest/Document/Multiple/Bulk/BulkOperation/BulkOperationsCollection.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	/// an equivalent of <see cref="List{T}.AddRange"/>. Because operations from Elasticsearch are executed in order none of the types in
+	/// System.Collection.Concurrent can't be used for this. We need to preserve insert order and exposed indexed index because <see cref="IBulkResponse.Items"/>
+	/// is ordered and lines up with <see cref="IBulkRequest.Operations"/> allowing one to zip the two together.
+	/// </summary>
+	/// <typeparam name="TOperation"></typeparam>
+	[ComVisible(false)]
+	public sealed class BulkOperationsCollection<TOperation> : IList<TOperation>, IList
+		where TOperation : IBulkOperation
+	{
+		private readonly object _lock = new object();
```
Comment (codebrain) [inline src/Nest/Document/Multiple/Bulk/BulkOperation/BulkOperationsCollection.cs]: `ReaderWriterLockSlim` ?

Comment (Mpdreamz) [reply src/Nest/Document/Multiple/Bulk/BulkOperation/BulkOperationsCollection.cs]: Discussed this, assumption is that this collection is write heavy and will not benefit from `ReadWriterLockSlim`

**Comment by Mpdreamz** [comment]
merged/ported to master

**Pull request merged**